### PR TITLE
Allow EPHEMERAL without explicit default expression

### LINKED ISSUE #35641: Allow EPHEMERAL column implicit default
ref #9436
Currently EPHEMERAL column requires an explicit default value provided on CREATE TABLE - we can allow implicit default there

### PR DESCRIPTION
**Pull request by yakov-olkhovskiy**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
closes #35641
Allow EPHEMERAL without explicit default expression

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Parsers/ParserCreateQuery.h [hunk 1]
```diff
@@ -185,15 +187,23 @@ bool IParserColumnDeclaration<NameParser>::parseImpl(Pos & pos, ASTPtr & node, E
...
     }
+    else
+    {
+        if (s_ephemeral.ignore(pos, expected))
+        {
+            default_specifier = Poco::toUpper(std::string{pos_before_specifier->begin, pos_before_specifier->end});
+            if (!expr_parser.parse(pos, default_expression, expected) && type)
+                default_expression = std::make_shared<ASTLiteral>(DataTypeFactory::instance().get(type)->getDefault());
+        }
+    }
```
Comment (antonio2368) [inline src/Parsers/ParserCreateQuery.h:206]:
```suggestion
    else if (s_ephemeral.ignore(pos, expected))
    {
        default_specifier = Poco::toUpper(std::string{pos_before_specifier->begin, pos_before_specifier->end});
        if (!expr_parser.parse(pos, default_expression, expected) && type)
            default_expression = std::make_shared<ASTLiteral>(DataTypeFactory::instance().get(type)->getDefault());
    }
```

**Comment by antonio2368** [comment]
Maybe add a test for it?

**Review by antonio2368 — APPROVED** [review]
LGTM

Split build needs to be fixed.

**Comment by yakov-olkhovskiy** [comment]
oh... well... it will be complicated now :)
apparently datatypes lib already depends on parsers lib, and easy fix will create circular reference which will not work for shared libraries
things rapidly escalated :)

File: src/Parsers/ParserCreateQuery.h [hunk 2]
```diff
@@ -185,15 +186,23 @@ bool IParserColumnDeclaration<NameParser>::parseImpl(Pos & pos, ASTPtr & node, E
...
     {
         default_specifier = Poco::toUpper(std::string{pos_before_specifier->begin, pos_before_specifier->end});
 
         /// should be followed by an expression
         if (!expr_parser.parse(pos, default_expression, expected))
             return false;
     }
+    else if (s_ephemeral.ignore(pos, expected))
+    {
+        if (s_ephemeral.ignore(pos, expected))
```
Comment (antonio2368) [inline src/Parsers/ParserCreateQuery.h:199]: this doesn't seem right

Comment (yakov-olkhovskiy) [reply src/Parsers/ParserCreateQuery.h:199]: yes, looks very strange, especially I have tested it and it worked :)
will check

Comment (yakov-olkhovskiy) [reply src/Parsers/ParserCreateQuery.h]: should be fixed now

**Comment by antonio2368** [comment]
Honestly it's a bit confusing with setting it to `Null` field when there is no default expression but it works.

Have you thought about setting the `default_expression` to correct value if the specifier is `EPHEMERAL` and `expression` is `nullptr` before each `if (col.default_expression)` check?

**Comment by yakov-olkhovskiy** [comment]
yeah, setting to NULL is confusing, but it's exactly why I can't set it to default from beginning - it requires datatypes module - and it creates circular reference

regarding second - not sure which `default_expression` you are talking about - if you mean just create `ASTPtr default_expression` externally - then there is no point - it will be same initialization through ternary expression or with dedicated `if`.

**Comment by antonio2368** [comment]
Yeah, there are multiple `default_expression`. Let's leave it as is, anything else would probably be even more confusing.

**Pull request merged**